modules/tensor_mechanics/src/materials/lagrangian/ComputeLagrangianObjectiveStress.C

```diff
@@ -43,88 +58,141 @@ ComputeLagrangianObjectiveStress::initQpStatefulProperties()
+    else
+      mooseError("Internal error: unsupported objective rate.");
   }
 }
 
-void
-ComputeLagrangianObjectiveStress::computeQpObjectiveUpdate()
+RankTwoTensor
+ComputeLagrangianObjectiveStress::objectiveUpdateTruesdell(const RankTwoTensor & dS)
 {
-  // Common to most/all models
+  // Get the kinematic tensor
+  RankTwoTensor dL = RankTwoTensor::Identity() - _inv_df[_qp];
 
-  // Small stress increment (really this all that needs to be defined)
-  RankTwoTensor dS = _small_stress[_qp] - _small_stress_old[_qp];
-  // Increment in the spatial velocity gradient
+  // Update the Cauchy stress
+  auto [S, Jinv] = advectStress(_cauchy_stress_old[_qp] + dS, dL);
+
+  // Get the appropriate tangent tensor
+  RankFourTensor U = stressAdvectionDerivative(S);
+  _cauchy_jacobian[_qp] = cauchyJacobian(Jinv, U);
+
+  return S;
+}
+
+RankTwoTensor
+ComputeLagrangianObjectiveStress::objectiveUpdateJaumann(const RankTwoTensor & dS)
+{
+  // Get the kinematic tensor
   RankTwoTensor dL = RankTwoTensor::Identity() - _inv_df[_qp];
+  RankTwoTensor dW = 0.5 * (dL - dL.transpose());
 
-  // Get the appropriate update tensor
-  RankFourTensor J;
-  switch (_rate)
-  {
-    case ObjectiveRate::Truesdell:
-      J = updateTensor(dL);
-      break;
-    case ObjectiveRate::Jaumann:
-      J = updateTensor(0.5 * (dL - dL.transpose()));
-      break;
-  }
+  // Update the Cauchy stress
+  auto [S, Jinv] = advectStress(_cauchy_stress_old[_qp] + dS, dW);
+
+  // Get the appropriate tangent tensor
+  RankTwoTensor I = RankTwoTensor::Identity();
+  RankFourTensor ddW_ddL = 0.5 * (I.mixedProductIkJl(I) - I.mixedProductIlJk(I));
+  RankFourTensor U = stressAdvectionDerivative(S) * ddW_ddL;
+  _cauchy_jacobian[_qp] = cauchyJacobian(Jinv, U);
+
+  return S;
+}
+
+RankTwoTensor
+ComputeLagrangianObjectiveStress::objectiveUpdateGreenNaghdi(const RankTwoTensor & dS)
+{
+  // The kinematic tensor for the Green-Naghdi rate is
+  // Omega = dot(R) R^T
+  polarDecomposition();
+  RankTwoTensor I = RankTwoTensor::Identity();
+  RankTwoTensor dR = (*_rotation)[_qp] * (*_rotation_old)[_qp].transpose() - I;
+  RankTwoTensor dO = dR * _inv_df[_qp];
 
   // Update the Cauchy stress
-  RankFourTensor Jinv = J.inverse();
-  _cauchy_stress[_qp] = Jinv * (_cauchy_stress_old[_qp] + dS);
+  auto [S, Jinv] = advectStress(_cauchy_stress_old[_qp] + dS, dO);
 
   // Get the appropriate tangent tensor
-  RankFourTensor U;
-  switch (_rate)
-  {
-    case ObjectiveRate::Truesdell:
-      U = truesdellTangent(_cauchy_stress[_qp]);
-      break;
-    case ObjectiveRate::Jaumann:
-      U = jaumannTangent(_cauchy_stress[_qp]);
-      break;
-  }
+  RankFourTensor d_R_d_F = (*_d_rotation_d_def_grad)[_qp];
+  RankFourTensor d_F_d_dL = _inv_df[_qp].inverse().mixedProductIkJl(_def_grad[_qp].transpose());
+  RankTwoTensor T = (*_rotation_old)[_qp].transpose() * _inv_df[_qp];
+  RankFourTensor d_dO_d_dL = T.mixedProductJkIjlm(d_R_d_F * d_F_d_dL) - dR.mixedProductIkJl(I);
```

Comment:
I'm getting lost on what depends on what, but can't we use @dschwen new einsum stuff to do these?